src/SACA.chpl

```diff
@@ -0,0 +1,207 @@
+    radixPass(SA12, s12, tmps, n02, K);
+    radixPass(s12, SA12, s , n02, K);
+    // find lexicographic names of triples
+    var name = 0:int, c0 = -1:int, c1 = -1:int, c2 = -1:int;
+
+    for i in 0..n02-1 do {
+      if (s[SA12[i]] != c0 || s[SA12[i]+1] != c1 || s[SA12[i]+2] != c2) {
+        name=name+1;
+        c0 = s[SA12[i]];
+        c1 = s[SA12[i]+1];
+        c2 = s[SA12[i]+2];
+      }
+      if (SA12[i] % 3 == 1) {
+        s12[SA12[i]/3] = name; // mapping the suffix to small alphabets
+      } // left half
+      else {
+        s12[SA12[i]/3 + n0] = name;
+      } // right half
+    }
+
+    // recurse if names are not unique
+    if (name < n02) {
+      SuffixArraySkew(s12, SA12, n02, name);
+      // store unique names in s12 using the suffix array
+      for i in 0..n02-1 do s12[SA12[i]] = i + 1;
+      //restore the value of s12 since we will change its values during the procedure
+    }
+    else { // generate the suffix array of s12 directly
+      for i in 0..n02-1 do SA12[s12[i] - 1] = i;
+      // here SA12 is in fact the ISA array.
+    }
+    // stably sort the mod 0 suffixes from SA12 by their first character
+    j=0;
+    for i in 0..n02-1 do {
+      // here in fact we take advantage of the sorted SA12 to just sort s0 once to get its sorted array
+      // at first we think the postion i%3=1 is the position
+      if (SA12[i] < n0) {
+        s0[j] = 3*SA12[i];
+        j=j+1;
+      }
+    }
+    radixPass(s0, SA0, s, n0, K);
+
+    // merge sorted SA0 suffixes and sorted SA12 suffixes
+    var p=0:int; // first s0 position
+    var t=n0-n1:int; //first s1 position
+    k=0;
+    var i1:int, j1:int;
+    var tmpk:int;
+    for tmpk in 0..n-1 do {
+      proc GetI():int {
+        if (SA12[t] < n0) {
+          return SA12[t] * 3 + 1;
+        }
+        else {
+          return (SA12[t]-n0) * 3 + 2;
+        }
+      }
+      i = GetI(); // pos of current offset 12 suffix
+      j = SA0[p]; // pos of current offset 0 suffix
+      var flag:bool;
+      if (SA12[t] < n0) {
+        // different compares for mod 1 and mod 2 suffixes
+        // i % 3 =1
+        flag=leq(s[i], s12[SA12[t]+n0], s[j], s12[j/3]);
+      }
+      else {
+        // i % 3 =2
+        flag=leq(s[i],s[i+1],s12[SA12[t]-n0+1], s[j], s[j+1], s12[j/3+n0]);
+      }
+      if (flag) {
+        // suffix from SA12 is smaller
+        SA[k] = i;
+        k=k+1;
+        t=t+1;
+        if (t == n02) {
+          // done --- only SA0 suffixes left
+          forall (i1,j1) in zip (k..n-1, p..p+n-k-1) do SA[i1] = SA0[j1];
+          break;
+        }
+      }
+      else {
+        // suffix from SA0 is smaller
+        SA[k] = j;
+        k=k+1;
+        p=p+1;
+        var tmpt=t:int;
+        if (p == n0) {
+          // done --- only SA12 suffixes left
+          for i1 in tmpt..n02-1 do {
+            SA[k] = GetI();
+            t=t+1;
+            k=k+1;
+          }
+          break;
+        }
+      }
+    }
+  }
+}
```

Comment:
General note: There is a lot going on in here, it would be nice to get
1. Chapel doc comments
2. Chapel based unit tests to illustrate proper function.

Reply:
@zhihuidu do you mind adding the chapel docs and unit tests for `SACA.chpl` since you're more familiar with this section than me?

Also if you could add the parenthesis to clarify the logic in this section like glitch suggested above while you're at it, that'd be great

```python
  inline proc leq(a1:int, a2:int, b1:int, b2:int) {
    // lexicographic order
    return(a1 < b1 || a1 == b1 && a2 <= b2);
  } // for pairs

  inline proc leq(a1:int, a2:int, a3:int, b1:int, b2:int, b3:int) {
    // lexicographic order
    return(a1 < b1 || a1 == b1 && leq(a2, a3, b2, b3));
  } // for pairs
```

You can pull down the latest in my `suffix_array` branch and make a pull request with these changes?

Reply:
I added `UnitTestSuffixArrayConstruction.chpl` with expected output in `UnitTestSuffixArrayConstruction.good`. This covers both the example from the paper and from client-side `suffix-array` docstring. 

I added docstrings and comments to the best of my ability using `Section 3` and `appendix A` of [Linear Work Suffix Array Construction](https://www.cs.helsinki.fi/u/tpkarkka/publications/jacm05-revised.pdf). These can certainly be improved and this section rewritten for clarity because it is still quite busy and difficult to follow.